src/native/http_connection.c

```diff
@@ -0,0 +1,256 @@
+
+    bool shutdown_requested;
+};
+
+static void s_on_http_conn_setup(struct aws_http_connection *connection, int error_code, void *user_data) {
+    struct http_jni_connection *http_jni_conn = (struct http_jni_connection *)user_data;
+    assert(http_jni_conn);
+
+    // Save the native pointer
+    http_jni_conn->native_http_conn = connection;
+
+    AWS_LOGF_DEBUG(AWS_LS_HTTP_CONNECTION, "Connected: http_connection: %p, err: %d", (void *)connection, error_code);
+
+    assert(s_http_connection.on_connection_complete);
+    assert(http_jni_conn->java_http_conn);
+    // Call the Java Object's "onComplete" callback
+    if (http_jni_conn->java_http_conn) {
+        JNIEnv *env = aws_jni_get_thread_env(http_jni_conn->jvm);
+        (*env)->CallVoidMethod(
+            env, http_jni_conn->java_http_conn, s_http_connection.on_connection_complete, error_code);
+    }
+}
+
+static void s_on_http_conn_shutdown(struct aws_http_connection *connection, int error_code, void *user_data) {
+    struct http_jni_connection *http_jni_conn = (struct http_jni_connection *)user_data;
+
+    AWS_LOGF_DEBUG(AWS_LS_HTTP_CONNECTION, "Shutdown: http_connection: %p, err: %d", (void *)connection, error_code);
+
+    // Call the Java Object's "onShutdown" callback
+    if (http_jni_conn->java_http_conn) {
+        JNIEnv *env = aws_jni_get_thread_env(http_jni_conn->jvm);
+        (*env)->CallVoidMethod(
+            env, http_jni_conn->java_http_conn, s_http_connection.on_connection_shutdown, error_code);
+    }
+}
+
+/**
+ * Create a new aws_http_request_options struct with default values
+ */
+JNIEXPORT long JNICALL Java_software_amazon_awssdk_crt_http_HttpConnection_httpConnectionNew(
+    JNIEnv *env,
+    jclass jni_class,
+    jobject http_conn_jobject,
+    jlong jni_client_bootstrap,
+    jlong jni_socket_options,
+    jlong jni_tls_ctx,
+    jstring jni_endpoint,
+    jshort jni_port) {
+
+    struct aws_client_bootstrap *client_bootstrap = (struct aws_client_bootstrap *)jni_client_bootstrap;
+    struct aws_socket_options *socket_options = (struct aws_socket_options *)jni_socket_options;
+    struct aws_tls_ctx *tls_ctx = (struct aws_tls_ctx *)jni_tls_ctx;
+
+    if (!client_bootstrap) {
+        aws_jni_throw_runtime_exception(env, "HttpConnection.httpConnectionNew: jni_client_bootstrap can't be null");
+        return (jlong)NULL;
+    }
+
+    if (!socket_options) {
+        aws_jni_throw_runtime_exception(env, "HttpConnection.httpConnectionNew: jni_socket_options can't be null");
+        return (jlong)NULL;
+    }
+
+    struct aws_allocator *allocator = aws_jni_get_allocator();
+    struct aws_byte_cursor endpoint = aws_jni_byte_cursor_from_jstring(env, jni_endpoint);
+    uint16_t port = jni_port;
+
+    if (port <= 0 || 65536 <= port) {
+        aws_jni_throw_runtime_exception(env, "HttpConnection.httpConnectionNew: Port must be between 1 and 65535");
+        return (jlong)NULL;
+    }
+
+    int use_tls = !(port == 80 || port == 8080);
```

Comment:
I also don't think this is a guarantee. Why not rely on the jni_tls_ctx being set or not?

Reply:
Updated Java side to require `http` or `https` Scheme for all URI's, and use the URI scheme to determine whether to use TLS or not. Java side then uses Scheme to decide whether or not to pass TlsContext/NULL to Native, and Native now always uses `TlsContext` if present.